src/hook/useNion.js

```diff
@@ -0,0 +1,213 @@
+                coercedDeclaration.dataKey
+            ],
+        }),
+        [selectNionResourcesForDataKeys, coercedDeclaration.dataKey],
+    )
+
+    // get entity, request, and extra data from the store
+    const { nion } = useMappedState(mapStateToProps)
+
+    // copy pasta from decorator
+    function getUrl(passedDeclaration, params) {
+        let endpoint = get(passedDeclaration, 'endpoint')
+        const buildUrl = ApiManager.getBuildUrl(passedDeclaration.apiType)
+
+        // If supplied endpoint override at call time, then use the supplied endpoint
+        if (get(params, 'endpoint')) {
+            endpoint = params.endpoint
+            params = omit(params, ['endpoint'])
+        }
+
+        // Use if a fully-formed url, otherwise pass to buildUrl
+        return typeof buildUrl === 'undefined' ||
+            typeof endpoint === 'undefined' ||
+            endpoint.indexOf('https://') === 0 ||
+            endpoint.indexOf('http://') === 0
+            ? endpoint
+            : buildUrl(endpoint, params)
+    }
+
+    const getResources = useCallback(
+        (params, actionOptions = {}) => {
+            const endpoint = getUrl(coercedDeclaration, params)
+            return nionActions.get(coercedDeclaration.dataKey, {
+                declaration: coercedDeclaration,
+                endpoint,
+                meta: {
+                    append: get(actionOptions, 'append'),
+                },
+            })(dispatch)
+        },
+        [coercedDeclaration],
+    )
+
+    const postResource = useCallback(
+        (body = {}, params, actionOptions) => {
+            const endpoint = getUrl(coercedDeclaration, params)
+
+            return nionActions.post(coercedDeclaration.dataKey, {
+                endpoint,
+                declaration: coercedDeclaration,
+                body,
+                meta: {
+                    append: get(actionOptions, 'append'),
+                },
+            })(dispatch)
+        },
+        [coercedDeclaration],
+    )
+
+    const patchResource = useCallback(
+        (body = {}, params) => {
+            const endpoint = getUrl(coercedDeclaration, params)
+            return nionActions.patch(coercedDeclaration.dataKey, {
+                endpoint,
+                declaration: coercedDeclaration,
+                body,
+            })(dispatch)
+        },
+        [coercedDeclaration],
+    )
+
+    const deleteDispatchFn = useCallback(
+        (refToDelete = {}, params, options = {}) => {
+            // TODO: Refactor ref to delete to not be mutative.
+            if (options.refToDelete) {
+                refToDelete = options.refToDelete
+            }
+            const endpoint = getUrl(coercedDeclaration, params)
+            return nionActions.delete(coercedDeclaration.dataKey, {
+                ...options,
+                declaration: coercedDeclaration,
+                endpoint,
+                refToDelete,
+            })(dispatch)
```

Comment:
i'm a little worried about drift between these functions and their equivalents in the decorator

Reply:
yea unfortunately i had a big refactor that abstracted the actions and interface so that we could provide multiple interfaces (component renderProp and container at the time)